# EJEMPLO METRICAS DE EVALUACIÓN

**Equipo:** 8

**Grupo:** 002 (Miercoles 7-10 pm)

**Integrantes:**

* 1872932 Huerta Torres Daniela Paola
* 1855188 Aguilar Ortiz Verónica Vanessa
* 1751125 Rodríguez Chaires Fernando Jahir
* 1857876 Jalomo Hernández Miguel Ángel


## BASE DE DATOS: Inicio de la diabetes de los indios Pima

Hay una gran cantidad de métricas que se pueden usar para evaluar las predicciones de estos problemas.

En esta sección revisaremos cómo utilizar las siguientes métricas:

* Precisión de clasificación.
* Pérdida de registros.
* Área bajo la curva ROC.
* Matriz de confusión.
* Informe de clasificación.

A continuación daremos a conocer la base de datos que estamos trabajando.

In [1]:
import pandas as pd
import numpy as np
import json

df=pd.read_csv("C:/Users/Daniela/Desktop/Daniela/Facultad/7mo sem/Mineria/pima-indians-diabetes.data.csv",encoding='utf-8')
print("Base de Datos:")
df

Base de Datos:


6  148  72  35    0  33.6  0.627  50  1
0     1   85  66  29    0  26.6  0.351  31  0
1     8  183  64   0    0  23.3  0.672  32  1
2     1   89  66  23   94  28.1  0.167  21  0
3     0  137  40  35  168  43.1  2.288  33  1
4     5  116  74   0    0  25.6  0.201  30  0
..   ..  ...  ..  ..  ...   ...    ...  .. ..
762  10  101  76  48  180  32.9  0.171  63  0
763   2  122  70  27    0  36.8  0.340  27  0
764   5  121  72  23  112  26.2  0.245  30  0
765   1  126  60   0    0  30.1  0.349  47  1
766   1   93  70  31    0  30.4  0.315  23  0

[767 rows x 9 columns]

Ahora continuamos con la precisión de la clasificación es el número de predicciones correctas realizadas como una proporción de todas las predicciones realizadas.

Esta es la métrica de evaluación más común para los problemas de clasificación, pero también es la que se utiliza con mayor frecuencia. En realidad, solo es adecuado cuando hay un número igual de observaciones en cada clase (lo que rara vez es el caso) y que todas las predicciones y errores de predicción son igualmente importantes, lo que a menudo no es el caso.

A continuación se muestra un ejemplo de cómo calcular la precisión de la clasificación.

In [3]:
import pandas
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = pandas.read_csv(url, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
kfold = model_selection.KFold(n_splits=10, random_state=7, shuffle=True)
model = LogisticRegression(solver='liblinear')
scoring = 'accuracy'
results = model_selection.cross_val_score(model, X, Y, cv=kfold, scoring=scoring)
print("Precisión: %.3f (%.3f)" % (results.mean(), results.std()))

Precisión: 0.771 (0.051)


Esto se puede convertir en un porcentaje multiplicando el valor por 100, lo que da una puntuación de precisión de aproximadamente un 77.1% de precisión.

La pérdida logística (o pérdida logarítmica) es una métrica de rendimiento para evaluar las predicciones de las probabilidades de pertenencia a una clase determinada.

La probabilidad escalar entre 0 y 1 puede verse como una medida de confianza para una predicción mediante un algoritmo. Las predicciones que son correctas o incorrectas son recompensadas o castigadas proporcionalmente a la confianza de la predicción.

A continuación se muestra un ejemplo del cálculo de la pérdida logarítmica para las predicciones de regresión logística en el conjunto de datos de inicio de la diabetes de los indios Pima.

In [5]:
import pandas
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = pandas.read_csv(url, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
kfold = model_selection.KFold(n_splits=10, random_state=7, shuffle=True)
model = LogisticRegression(solver='liblinear')
scoring = 'neg_log_loss'
results = model_selection.cross_val_score(model, X, Y, cv=kfold, scoring=scoring)
print("Perdida de registros: %.3f (%.3f)" % (results.mean(), results.std()))

Perdida de registros: -0.494 (0.042)


Una pérdida de logaritmo más pequeña es mejor si 0 representa una pérdida de logaritmo perfecta.



El área bajo la curva ROC (o ROC AUC para abreviar) es una métrica de rendimiento para problemas de clasificación binaria.

El AUC representa la capacidad de un modelo para discriminar entre clases positivas y negativas. Un área de 1.0 representa un modelo que hizo todas las predicciones a la perfección. Un área de 0.5 representa un modelo tan bueno como aleatorio.

Una curva ROC es un gráfico de la tasa de verdaderos positivos y la tasa de falsos positivos para un conjunto dado de predicciones de probabilidad en diferentes umbrales que se utilizan para asignar las probabilidades a las etiquetas de clase. El área bajo la curva es entonces la integral aproximada bajo la Curva ROC.

A continuación se realiza el cálculo de AUC

In [14]:
import pandas
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = pandas.read_csv(url, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
kfold = model_selection.KFold(n_splits=10, random_state=7, shuffle=True)
model = LogisticRegression(solver='liblinear')
scoring = 'roc_auc'
results = model_selection.cross_val_score(model, X, Y, cv=kfold, scoring=scoring)
print("AUC: %.3f (%.3f)" % (results.mean(), results.std()))


AUC: 0.826 (0.050)


Cálculo de una matriz de confusión para un conjunto de predicciones mediante un modelo en un conjunto de prueba.

La tabla presenta predicciones en el eje xy resultados de precisión en el eje y. Las celdas de la tabla son el número de predicciones realizadas por un algoritmo de aprendizaje automático.

In [6]:
import pandas
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = pandas.read_csv(url, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
test_size = 0.33
X_train, X_test, Y_train, Y_test = model_selection.train_test_split(X, Y, test_size=test_size, random_state=7)
model = LogisticRegression(solver='liblinear')
model.fit(X_train, Y_train)
predicted = model.predict(X_test)
matrix = confusion_matrix(Y_test, predicted)
print(matrix)

[[141  21]
 [ 41  51]]


La libreria scikit-learn proporciona un informe práctico cuando se trabaja en problemas de clasificación para darle una idea rápida de la precisión de un modelo utilizando una serie de medidas.

La función class_report () muestra la precisión, recuperación, puntaje f1 y soporte para cada clase.

El siguiente ejemplo muestra el informe sobre el problema de clasificación binaria.

In [16]:
import pandas
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = pandas.read_csv(url, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
test_size = 0.33
X_train, X_test, Y_train, Y_test = model_selection.train_test_split(X, Y, test_size=test_size, random_state=7)
model = LogisticRegression(solver='liblinear')
model.fit(X_train, Y_train)
predicted = model.predict(X_test)
report = classification_report(Y_test, predicted)
print(report)

              precision    recall  f1-score   support

         0.0       0.77      0.87      0.82       162
         1.0       0.71      0.55      0.62        92

    accuracy                           0.76       254
   macro avg       0.74      0.71      0.72       254
weighted avg       0.75      0.76      0.75       254

